### ISSUE TITLE
can't re-establish severed gossip connection until the old connection times out

### ISSUE THREAD

**Issue by arilotter** [n0-computer/iroh-gossip#10]
I have some program that panics while connected to gossip.
When i restart it, and get it to connect to one of its old peers it used to be connected to, i see
`ignoring connection: already accepted peer_id=PublicKey(…) origin=Accept.`
Meanwhile, the node attempting to reconnect panics the task with the error: `connection closed: connection lost.`

This appears to happen because Gossip only wants to maintain one connection per pair, and I'd need to wait for the existing (broken) connection to time out.

It would be great to have an option to forcibly replace any existing connection when sending `Command::JoinPeers` from a node.

**Comment by matheus23** [comment]
I've recently been thinking about this issue again.

At the moment, if we already have a running connection that was `Endpoint::accept()`ed, then we reject any new ones:
https://github.com/n0-computer/iroh-gossip/blob/534f01046332a21f6356d189c686f7c6c17af3c2/src/net.rs#L941-L943

This is what's causing the issue that stale connections first need to time out, before they can be replaced with fresh ones, *if* the stale connection was accepted, instead of initiated by ourselves.

There's a similar mechanism for maintaining connections in iroh-willow, it has similar behaviour:
https://github.com/n0-computer/iroh-willow/blob/bc7468f63ae3713f8ff9a753c5b4ea75044f5aca/src/engine/peer_manager.rs#L247-L251

However, it additionally looks at deduplicating connections where you have both dialed but also accepted a connection from the same peer, using a [comparison on the `node_id`](https://github.com/n0-computer/iroh-willow/blob/bc7468f63ae3713f8ff9a753c5b4ea75044f5aca/src/engine/peer_manager.rs#L236) to find out which connection should stay alive.

Instead I think a good connection manager should probably do the following:
1. Always accept incoming connections. It won't guarantee that it will only ever have one connection. So sometimes you'll have to duplicate your work on multiple connections to the same peer (you'd probably want to have a "multiple redundant connection limit" of 5 or something like that).
2. When you accept a connection from a peer that you have an *initiated* connection with, then do the comparison trick to decide whether to close the initiated connection. If you lose the comparison, you'll keep both connections, and need to trust the other side to do the same comparison trick to close their connection.
3. If application code makes you initiate a second connection, deduplicate that one. It'd probably be wise to have two flavors of "get me a connection", one which closes all previous (initiated!) connections and replaces them with a new one, and one that gives you the "best" connection you already have initiated.

The core principle here is essentially "you're in charge of the connections you *initiate*, and the other side is in charge of theirs" (i.e. you're not in charge of closing/managing connections you've accepted).

I think it should be feasible to write a test for such a connection manager, where we simulate a network with latency and occasional hard resets for nodes, and make sure that we sufficiently deduplicate connections in these cases, but also don't stall messages for long periods of times where we just wait for connections to time out.

**Comment by johnny-smitherson** [comment]
This bug really hurts reliability when a quick disconnect & reconnect is turned into a forced timeout of minutes

So, basically, you can't use the same node id when reconnecting. So you must randomize the node id every time you join a gossip, or when you detect a disconnect. Since the node id is now random, it needs to be discoverable.

This means you now need to run a second endpoint, on your saved/"permanent" node id, that never connects to gossip. This endpoint only runs some echo protocol that returns the current randomized gossip id, so other nodes know how to reconnect to your tickets, if they happen to restart/reconnect too.

This is getting pretty complex on the application side, which hurts reliability even more.


> This is what's causing the issue that stale connections first need to time out, before they can be replaced with fresh ones, if the stale connection was accepted, instead of initiated by ourselves.

What if, instead of having 1 connection per peer, you have 2 - one outgoing and one incoming. 

They could be single-direction instead of bidirectional. 

Each peer would retry its outgoing connection in case of errors or disconnect. Incoming connections would always drop the previous one from that specific sender -- they're only opening a second one because the first one is failing, right?

This might be simpler to implement and test

**Comment by Frando** [comment]
Finally had the time to look into this properly, see #43. I'd be glad for someone testing this in a real-world scenario to make sure it *actually* works outside of test scenarios.

**Comment by johnny-smitherson** [comment]
Yes your branch fixes the chat where some nodes have static IDs.

Here is our instance https://sparganothis.org/Sparganothis-v2/chat - nodes tagged with `bootstrap #i` have const IDs. 

Without your patch, they cannot reconnect to the swarm after user refreshes page.

Thanks!
